daemon/internal/newrelic/php_packages.go

```diff
@@ -34,51 +34,112 @@ func (packages *PhpPackages) NumSaved() float64 {
 // newPhpPackages returns a new PhpPackages struct.
 func NewPhpPackages() *PhpPackages {
 	p := &PhpPackages{
-		numSeen: 0,
-		data:    nil,
+		numSeen:      0,
+		data:         nil,
+		filteredPkgs: nil,
 	}
 
 	return p
 }
 
-// SetPhpPackages sets the observed package list.
-func (packages *PhpPackages) SetPhpPackages(data []byte) error {
+// filter seen php packages data to avoid sending duplicates
+//
+// the `App` structure contains a map of PHP Packages the reporting
+// application has encountered.
+//
+// the map of packages should persist for the duration of the
+// current connection
+//
+// the JSON format received from the agent is:
+//
+//	[["package_name","version",{}],...]
+//
+// for each entry, assign the package name and version to the `PhpPackagesKey`
+// struct and use the key to verify data does not exist in the map. If the
+// key does not exist, add it to the map and the slice of filteredPkgs to be
+// sent in the current harvest.
+func (packages *PhpPackages) Filter(pkgHistory map[PhpPackagesKey]struct{}) {
+	if packages.data == nil {
+		return
+	}
 
-	if nil == packages {
-		return fmt.Errorf("packages is nil!")
+	for _, pkgKey := range packages.data {
+		_, ok := pkgHistory[pkgKey]
+		if !ok {
+			pkgHistory[pkgKey] = struct{}{}
```

Comment:
Why an empty struct?

Reply:
We store an empty struct as the value to indicate that an index has been hit, since we don't care about the actual value but instead the existence of the package key in the map. This is a trick in Go to populate a map where you are only concerned with the existence of a key.